assert: improve assert.throws

BridgeAR opened the pull request:
Throw a TypeError in case a error message is provided in the second
argument and a third argument is present as well.
This is clearly a mistake and should not be done.

CI https://ci.nodejs.org/job/node-test-pull-request/12015/

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
assert

Trott commented on doc/api/assert.md:
````diff
@@ -772,7 +772,7 @@ argument, then `error` is assumed to be omitted and the string will be used for
 ```js
 // THIS IS A MISTAKE! DO NOT DO THIS!
-assert.throws(myFunction, 'missing foo', 'did not throw with expected message');
+assert.throws(myFunction, 'missing foo'); // Did not throw with expected message
````
The text above this example needs to be re-written now as well as the example on line 778 as well. Rather than moving the third argument to a comment, perhaps just remove it entirely. The user should still be cautioned against using a string for the expected error message because a string will be interpreted as the message to provide if the function doesn't throw rather than something to validate the thrown error.

BridgeAR replied:
I am not sure I can follow. The situation was only improved in case three arguments are passed to `throws`. Otherwise the problem continues to exist. So the mitigation is only very tiny as the typical wrong use case is probably `assert.throws(fn, "error msg")`.

I thought a few times about how to improve the two args situation but as far as I see it there is nothing that can be done to improve it. I actually also thought about deprecating using a string as second argument but that would be very inconsistent and weird as well.

Trott replied:
Refining my comment a bit: The current version of the example code (`assert.throws(fn, str, str)`) is obviously an error. The version above in this PR (`assert.throws(fn, str)`) is less obviously an error because maybe "missing foo" is not an expected error message but rather the error message we want to provide if the function doesn't throw as expected. It's not clear to me if the best way to clarify is to edit the text preceding the block or if it is to edit `missing foo` to say something else.

BridgeAR commented:
I updated the example and tried to be as detailed and precise as possible.

BridgeAR commented:
While thinking about this further: is this actually a semver-patch instead of semver-major? Because the use case when a error is now thrown is clearly a mistake and did not work as the user intended.

Trott commented on doc/api/assert.md:
````diff
@@ -767,17 +767,39 @@ assert.throws(
 ```
 
+Due to the confusing notation it is recommended not to use a string as second
+argument at all. This might otherwise lead to difficult to spot errors.
````
Nit: `difficult to spot` -> `difficult-to-spot`

Trott commented on doc/api/assert.md:
````diff
@@ -767,17 +767,39 @@ assert.throws(
+// AssertionError [ERR_ASSERTION]: Error: Mismatch
 ```
 
+Due to the confusing notation it is recommended not to use a string as second
````
Should it be `as the second` instead of `as second`? That seems more natural to me.

BridgeAR replied:
Hm, for me `as second` sounds better but I am probably not the best one to ask in this case. Are there any other votes? :laughing:

Trott commented on doc/api/assert.md:
```diff
@@ -767,17 +767,39 @@ assert.throws(
+// If you originally intended to match for the error message do this instead:
+assert.throws(throwingMatched, /^Error: Matched$/);
+// Does not throw because the error messages match.
+assert.throws(throwingMismatch, /^Error: Matched$/);
```
Am I missing something subtle? It looks like line 794 and line 796 are the same?

BridgeAR replied:
The called function is named similar but different. I clarified it now by choosing a different name.

Trott commented:
@BridgeAR I'm OK with patch for this but would be interested in some other @nodejs/tsc opinions. Technically, anything that introduces a `throw` in a situation where it previously was not throwing is supposed to be treated as a breaking change, but I'm not sure that really applies in a situation like this where the code that doesn't throw is kinda broken anyway.

Trott commented:
Left some nits on the new doc changes, but I really like the direction you went with it. 👍

BridgeAR commented:
I updated it and also added a independent commit to clarify how the RegExp actually works.

mcollina commented:
I would release this as minor. Mainly to give it a bit more time to bake for LTS.
I'm fine with patch as well.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-commit-light/18/

jasnell commented on doc/api/assert.md:
```diff
@@ -767,17 +770,41 @@ assert.throws(
+assert.throws(notThrowing, 'Second');
+// AssertionError [ERR_ASSERTION]: Missing expected exception: Second
+
+// If you originally intended to match for the error message do this instead:
```
Please void using informal pronouns like `you` in the docs :-)

BridgeAR replied:
I rephrased that but I am not really happy with what I came up with
`Please read the example below carefully if using a string as the second argument gets considered:`
and
`If it was intended to match for the error message do this instead:`

maclover7 commented on doc/api/assert.md:
````diff
@@ -767,17 +770,41 @@ assert.throws(
+//     at throwingFirst (repl:2:9)
 ```
 
+Due to the confusing notation it is recommended not to use a string as second
````
- `Comma after notation`
- `as a second argument`
- ✂️ `at all`
- ✂️ `otherwise`

maclover7 commented on doc/api/assert.md:
```diff
@@ -740,12 +740,15 @@ assert.throws(
 
 Validate error message using [`RegExp`][]:
 
+Using a regular expression runs `.toString` on the error object and will
```
Needs a comma after object

maclover7 commented on doc/api/assert.md:
```diff
@@ -767,17 +770,41 @@ assert.throws(
 argument, then `error` is assumed to be omitted and the string will be used for
-`message` instead. This can lead to easy-to-miss mistakes:
+`message` instead. This can lead to easy-to-miss mistakes. Please read the
+example below carefully if you intend to use a string as second argument:
```
`as a second argument`

evanlucas replied:
I would think it should be more like `as the second argument` since there can only be one

BridgeAR commented:
I addressed the comments but I am not that happy with all the changes. So it would be nice to get another look at the sentences that contained a `you` before.

BridgeAR commented:
New CI https://ci.nodejs.org/job/node-test-commit-light/30/

maclover7 commented:
@BridgeAR I think this should be using `node-test-pull-request` not `node-test-commit-light`, in order to run the tests on all platforms. `node-test-commit-light` is usually only meant to be used on documentation only PRs

BridgeAR commented:
@maclover7 I intentionally used that as it should only verify that the rebase was fine. `assert` has no special behavior on any platform and it should always be sufficient to only run `linux-one` in that case.

It takes ages to run a full CI for changes that absolutely behave the same on all platforms and there is no point in doing that out of my perspective.

BridgeAR commented:
PTAL

BridgeAR commented:
Ping @Trott @mcollina @maclover7 @jasnell @evanlucas @benjamingr @gireeshpunathil

mcollina commented:
Still LGTM but it needs a rebase.

BridgeAR commented:
To many conflicting PRs... rebased.

BridgeAR commented:
Landed in fe257ac2078676012ed6d188e63e50827fe69d27, da5c7d68cdceaa70411ffab6bee16e200a703aa9

(The only conflict was the stackFrameFunction and I ran the tests again locally before pushing)

MylesBorins commented:
This does not land cleanly on v9.x, would someone be willing to backport?

BridgeAR commented:
Backport opened in #19230

MylesBorins commented:
Requested backport to 8.x in #19230
